### PR TITLE
Switch importers to the new input pipeline

### PR DESCRIPTION
**Pull request by reuben**
I haven't finished testing with datasets other than LDC93S1 yet, and I still need to figure out a proper fix for the race between importer and training, but since this work is blocking the work week I'm publishing it anyway. Maybe it's best to merge it with the hack to unblock others and then fix it.

### CODE REVIEW

**Comment by reuben** [comment]
I wasn't particularly careful with splitting these changes into contained commits as I wanted to share them as soon as I could, but I can do that.

File: DeepSpeech.ipynb [hunk 1]
```diff
@@ -1415,12 +1405,20 @@
...
+    "    # Start queue runner threads\n",
+    "    tf.train.start_queue_runners(sess=session)\n",
+    "    \n",
+    "    # Obtain all the data, defaulting to TED LIUM\n",
+    "    data_sets = ds_importer_module.read_data_sets(session, ds_dataset_path, batch_size, n_input, n_context)\n",
+    "    \n",
+    "    #TODO: there's a race between the importer's background threads feeding the example queue and the training code\n",
+    "    #      calling next_batch. AFAICT this should never happen as that's precisely what TF queues are supposed to\n",
+    "    #      handle. for now, just wait a while before starting training so we can fill the example queue.\n",
+    "    time.sleep(30)\n",
```
Comment (Cwiiis) [inline DeepSpeech.ipynb]: 30 seconds seems like an awfully long time... If we have to wait (I assume due to a bug, as like you say, this is the entire point of queues...), could we not include some kind of wait/yield to do it based on when the queue actually fills with enough data for the importers?

File: util/text.py [hunk 2]
```diff
@@ -121,3 +111,46 @@ def levenshtein(a,b):
             current[j] = min(add, delete, change)
 
     return current[n]
+
+# gather_nd is taken from https://github.com/tensorflow/tensorflow/issues/206#issuecomment-229678962
+# 
+# Unfortunately we can't just use tf.gather_nd because it does not have gradients
+# implemented yet, so we need this workaround.
+#
+def gather_nd(params, indices, shape):
```
Comment (kdavis-mozilla) [inline util/text.py]: For the time being I'm going to assume this is implemented correctly

File: util/text.py [hunk 3]
```diff
@@ -121,3 +111,46 @@ def levenshtein(a,b):
...
+    flat_indices = sum([a*b for a,b in zip(multipliers, indices_unpacked)])
+    return tf.gather(flat_params, flat_indices)
+
+# ctc_label_dense_to_sparse is taken from https://github.com/tensorflow/tensorflow/issues/1742#issuecomment-205291527
+#
+# The CTC implementation in TensorFlow needs labels in a sparse representation,
+# but sparse data and queues don't mix well, so we store padded tensors in the
+# queue and convert to a sparse representation after dequeuing a batch.
+#
+def ctc_label_dense_to_sparse(labels, label_lengths, batch_size):
```
Comment (kdavis-mozilla) [inline util/text.py]: For the time being I'm going to assume this is implemented correctly

**Comment by reuben** [comment]
I've been running a test on the server with the TED data for a couple of hours now and it's been going longer than it ever did. Since this PR is blocking people's work I suggest we merge it sooner rather than later. @kdavis-mozilla, can you review my subsequent commits?

**Comment by reuben** [comment]
Epoch 1 completed :)

https://gist.github.com/reuben/2e46a88336ecf1d38dfee175dd8e1715

**Comment by kdavis-mozilla** [comment]
LGTM

**Pull request merged**